library/common/engine.h

```diff
@@ -48,12 +48,12 @@ class Engine {
   envoy_engine_callbacks callbacks_;
   Thread::MutexBasicLockable mutex_;
   Thread::CondVar cv_;
-  std::thread main_thread_;
   std::unique_ptr<Http::Dispatcher> http_dispatcher_;
   std::unique_ptr<MobileMainCommon> main_common_ GUARDED_BY(mutex_);
   Server::Instance* server_{};
   Server::ServerLifecycleNotifier::HandlePtr postinit_callback_handler_;
   Event::Dispatcher* event_dispatcher_;
+  std::thread main_thread_;
```

Comment:
Can we document this inline somewhere? We're now relying on implementation details of how destructors are called, which is very fragile. Tests would be ideal, but I understand this is probably quite difficult to test, so we should at least document it

Reply:
> We're now relying on implementation details of how destructors are called, which is very fragile

For better or worse this is a common c++ pattern.

> Tests would be ideal

As always agreed. I had the idea for this crash "fix" but I am not entirely sure that the situation described here with instructions on the main_thread being executed after destruction of other objects is the actual problem. I tried writing repro tests but was not able to set it up. I have a related PR with engine tests coming up imminently that repros a different trace. I can try an massage that to repro this as well.

> Can we document this inline somewhere

Absolutely, this should be in the comments. Updated.